engine/baml-compiler/src/lib.rs

```diff
@@ -759,6 +824,109 @@ mod tests {
         })
     }
 
+    #[test]
+    fn create_iterator_instruction() -> anyhow::Result<()> {
+        use baml_vm::{Bytecode, Frame, Function, FunctionKind, Instruction, Object, Value, Vm};
+
+        // Create a simple function that tests CreateIterator instruction
+        let function = Function {
+            name: "test_create_iterator".to_string(),
+            arity: 0,
+            bytecode: Bytecode {
+                instructions: vec![
+                    Instruction::LoadConst(0),   // Load 1
+                    Instruction::LoadConst(1),   // Load 2
+                    Instruction::LoadConst(2),   // Load 3
+                    Instruction::AllocArray(3),  // Create array [1, 2, 3]
+                    Instruction::CreateIterator, // Create iterator
+                    Instruction::Return,
+                ],
+                constants: vec![Value::Int(1), Value::Int(2), Value::Int(3)],
+            },
+            kind: FunctionKind::Exec,
+            local_var_names: vec!["<fn test_create_iterator>".to_string()],
+        };
+
+        let objects = vec![Object::Function(function)];
+        let globals = vec![Value::Object(0)];
+
+        let mut vm = Vm {
+            frames: vec![Frame {
+                function: 0,
+                instruction_ptr: 0,
+                locals_offset: 0,
+            }],
+            stack: vec![Value::Object(0)],
+            objects,
+            globals,
+        };
+
+        let result = vm.exec()?;
+        // Result should be an iterator object
+        if let Value::Object(index) = result {
+            if let Object::Iterator {
+                array,
+                index: iter_index,
+            } = &vm.objects[index]
+            {
+                assert_eq!(*iter_index, 0); // Should start at index 0
+                                            // Array should be at index 1 (after function at index 0)
+                assert_eq!(*array, 1);
+            } else {
+                panic!("Expected Iterator object");
+            }
+        } else {
+            panic!("Expected Object value");
+        }
```

Comment:
When you have this big if blocks with a tiny else statement at the end, use Rust's `let-else` construct:

```rust
let Value::Object(index) = result else {
    panic!("Expected Object value");
};

let Object::Iterator { array, index: iter_index } = &vm.objects[index] else {
    panic!("Expected Iterator object");
};

assert_eq(*iter_index, 0);
assert_eq(*array, 1);
```

That way it's much easier to read each case and you don't have a lot of nesting. It's similar to early returns in functions.